man/cauthtool.8
.TH CAUTHTOOL 8
.SH NAME
cauthtool \- ceph keyring manipulation tool
.SH SYNOPSIS
.B cauthtool
\fIkeyringfile\fP
[ \fB\-l\fR | \fB\-\-list\fR ]
[ \fB\-C\fR | \fB\-\-create-keyring\fR ]
[ \fB\-p\fR | \fB\-\-print\fR ]
[ \fB\-n\fR | \fB\-\-name\fR \fIentityname\fP ]
[ \fB\-\-gen-key\fR ]
[ \fB\-a\fR | \fB\-\-add-key \fIbase64_key\fP ]
[ \fB\-\-caps\fR \fIcapfils\fP ]
[ \fB\-b\fR | \fB\-\-bin\fR ]
.SH DESCRIPTION
.B cauthtool
is a utility to create, view, and modify a Ceph keyring file.  A keyring
file stores one or more Ceph authentication keys and possibly an 
associated capability specification.  Each key is associated with an
entity name, of the form \fI{client,mon,mds,osd}.name\fP.
.SH OPTIONS
.TP
\fB\-l\fP, \fB\-\-list\fP
will list all keys and capabilities present in the keyring
.TP
\fB\-p\fP, \fB\-\-print\fP
will print an encoded key for the specified \fIentityname\fP.  This is suitable for the mount -o secret= argument
.TP
\fB\-C\fP, \fB\-\-create-keyring\fP
will create a new keyring, overwriting any existing \fIkeyringfile\fP
.TP
\fB\-\-gen\-key\fP
will generate a new secret key for the specified \fIentityname\fP
.TP
\fB\-\-add\-key\fP
will add an encoded key to the keyring
.TP
\fB\-\-cap\fI subsystem capability \fP
will set the capability for given subsystem
.TP
\fB\-\-caps\fI capsfile \fP
will set all of capabilities associated with a given key, for all subsystems
.TP
\fB\-b\fP, \fB\-\-bin\fP
will create a binary formatted keyring

.SH CAPABILITIES

The \fIsubsystem\fP is the name of a Ceph subsystem: mon, mds, or osd.
.PP
The \fIcapability\fP is a string describing what the given user is
allowed to do.  This takes the form of a comma separated list of
allow, deny clauses with a permission specifier containing one or more
of \fIrwx\fP for read, write, and execute permission.  The "allow *" grants
full superuser permissions for the given subsystem.
.PP
For example,

.IP
osd = "allow rwx [pool=foo[,bar]]|[uid=baz[,bay]]"  # can read, write, and execute objects
.IP
mds = "allow"      # can access mds server
.IP
mon = "allow rwx"  # can modify cluster state (i.e., is a server daemon)
.PP
A librados user restricted to a single pool might look like
.IP
osd = "allow rw pool foo"
.PP
A client mounting the file system with minimal permissions would need caps like
.IP
mds = "allow"
.IP
osd = "allow rw pool=data"
.IP
mon = "allow r"
.PP

.SH CAPS FILE FORMAT

The caps file format consists of zero or more key/value pairs, one per line.  The key and value are separated by an '=', and the value must be quoted (with ' or ") if it contains any whitespace.  The key is the name of the Ceph
subsystem (osd, mds, mon), and the value is the capability string (see above). 

.SH EXAMPLE
To create a new keyring containing a key for \fIclient.foo\fP:
.IP
cauthtool -c -n client.foo --gen-key keyring
.PP
To associate some capabilities with the key (namely, the ability to mount a Ceph filesystem):
.IP
cauthtool -n client.foo --cap mds 'allow' --cap osd 'allow rw pool=data' --cap mon 'allow r' keyring
.PP
To display the contents of the keyring:
.IP
cauthtool -l keyring
.PP
When mount a Ceph file system, you can grab the appropriately encoded secret key with
.IP
mount -t ceph serverhost:/ mountpoint -o name=foo,secret=`cauthtool -p -n client.foo keyring`
.PP
.SH AVAILABILITY
.B cauthtool
is part of the Ceph distributed file system.  Please refer to the Ceph wiki at
http://ceph.newdream.net/wiki for more information.
.SH SEE ALSO
.BR ceph (8)


src/cauthtool.cc
// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
// vim: ts=8 sw=2 smarttab
/*
 * Ceph - scalable distributed file system
 *
 * Copyright (C) 2004-2009 Sage Weil <sage@newdream.net>
 *
 * This is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1, as published by the Free Software 
 * Foundation.  See file COPYING.
 * 
 */

using namespace std;

#include "common/config.h"

#include "common/ConfUtils.h"
#include "common/ceph_argparse.h"
#include "common/common_init.h"
#include "auth/Crypto.h"
#include "auth/Auth.h"
#include "auth/KeyRing.h"

#include <sstream>

void usage()
{
  cout << "usage: cauthtool keyringfile [OPTIONS]...\n"
       << "where the options are:\n"
       << "  -l, --list                    will list all keys and capabilities present in\n"
       << "                                the keyring\n"
       << "  -p, --print                   will print an encoded key for the specified\n"
       << "                                entityname. This is suitable for the\n"
       << "                                'mount -o secret=..' argument\n"
       << "  -C, --create-keyring          will create a new keyring, overwriting any\n"
       << "                                existing keyringfile\n"
       << "  --gen-key                     will generate a new secret key for the\n"
       << "                                specified entityname\n"
       << "  --add-key                     will add an encoded key to the keyring\n"
       << "  --cap subsystem capability    will set the capability for given subsystem\n"
       << "  --caps capsfile               will set all of capabilities associated with a\n"
       << "                                given key, for all subsystems\n"
       << "  -b, --bin                     will create a binary formatted keyring" << std::endl;
  exit(1);
}

int main(int argc, const char **argv)
{
  vector<const char*> args;
  argv_to_vec(argc, argv, args);
  env_to_vec(args);
  DEFINE_CONF_VARS(usage);

  common_init(args, CEPH_ENTITY_TYPE_CLIENT, CODE_ENVIRONMENT_UTILITY,
	      CINIT_FLAG_NO_DEFAULT_CONFIG_FILE);
  EntityName ename(g_conf.name);

  const char *me = argv[0];

  const char *fn = 0;
  bool gen_key = false;
  bool gen_print_key = false;
  const char *add_key = 0;
  bool list = false;
  bool print_key = false;
  bool create_keyring = false;
  const char *caps_fn = NULL;
  const char *import_keyring = NULL;
  bool set_auid = false;
  uint64_t auid = CEPH_AUTH_UID_DEFAULT;
  map<string,bufferlist> caps;
  bool bin_keyring = false;

  FOR_EACH_ARG(args) {
    if (CEPH_ARGPARSE_EQ("gen-key", 'g')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&gen_key, OPT_BOOL);
    } else if (CEPH_ARGPARSE_EQ("gen-print-key", '\0')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&gen_print_key, OPT_BOOL);
    } else if (CEPH_ARGPARSE_EQ("add-key", 'a')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&add_key, OPT_STR);
    } else if (CEPH_ARGPARSE_EQ("list", 'l')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&list, OPT_BOOL);
    } else if (CEPH_ARGPARSE_EQ("caps", '\0')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&caps_fn, OPT_STR);
    } else if (CEPH_ARGPARSE_EQ("cap", '\0')) {
      const char *key, *val;
      CEPH_ARGPARSE_SET_ARG_VAL(&key, OPT_STR);
      CEPH_ARGPARSE_SET_ARG_VAL(&val, OPT_STR);
      ::encode(val, caps[key]);
    } else if (CEPH_ARGPARSE_EQ("print-key", 'p')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&print_key, OPT_BOOL);
    } else if (CEPH_ARGPARSE_EQ("create-keyring", '\0')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&create_keyring, OPT_BOOL);
    } else if (CEPH_ARGPARSE_EQ("import-keyring", '\0')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&import_keyring, OPT_STR);
    } else if (CEPH_ARGPARSE_EQ("set-uid", 'u')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&auid, OPT_LONGLONG);
      set_auid = true;
    } else if (CEPH_ARGPARSE_EQ("bin", 'b')) {
      CEPH_ARGPARSE_SET_ARG_VAL(&bin_keyring, OPT_BOOL);
    } else if (!fn) {
      fn = args[i];
    } else 
      usage();
  }
  if (!fn && !gen_print_key) {
    cerr << me << ": must specify filename" << std::endl;
    usage();
  }
  if (!(gen_key ||
	gen_print_key ||
	add_key ||
	list ||
	caps_fn ||
	caps.size() ||
	set_auid ||
	print_key ||
	create_keyring ||
	import_keyring)) {
    cerr << "no command specified" << std::endl;
    usage();
  }
  if (gen_key && add_key) {
    cerr << "can't both gen_key and add_key" << std::endl;
    usage();
  }	

  if (gen_print_key) {
    CryptoKey key;
    key.create(CEPH_CRYPTO_AES);
    cout << key << std::endl;    
    return 0;
  }

  // keyring --------
  bool modified = false;
  KeyRing keyring;

  bufferlist bl;
  int r = 0;
  if (create_keyring) {
    cout << "creating " << fn << std::endl;
    modified = true;
  } else {
    r = bl.read_file(fn, true);
    if (r >= 0) {
      try {
	bufferlist::iterator iter = bl.begin();
	::decode(keyring, iter);
      } catch (const buffer::error &err) {
	cerr << "error reading file " << fn << std::endl;
	exit(1);
      }
    } else {
      cerr << "can't open " << fn << ": " << strerror(-r) << std::endl;
      exit(1);
    }
  }

  // write commands
  if (import_keyring) {
    KeyRing other;
    bufferlist obl;
    int r = obl.read_file(import_keyring);
    if (r >= 0) {
      try {
	bufferlist::iterator iter = obl.begin();
	::decode(other, iter);
      } catch (const buffer::error &err) {
	cerr << "error reading file " << import_keyring << std::endl;
	exit(1);
      }
      
      cout << "importing contents of " << import_keyring << " into " << fn << std::endl;
      //other.print(cout);
      keyring.import(other);
      modified = true;
    } else {
      cerr << "can't open " << import_keyring << ": " << strerror(-r) << std::endl;
      exit(1);
    }
  }
  if (gen_key) {
    EntityAuth eauth;
    eauth.key.create(CEPH_CRYPTO_AES);
    keyring.add(ename, eauth);
    modified = true;
  }
  if (add_key) {
    EntityAuth eauth;
    string ekey(add_key);
    try {
      eauth.key.decode_base64(ekey);
    } catch (const buffer::error &err) {
      cerr << "can't decode key '" << add_key << "'" << std::endl;
      exit(1);
    }
    keyring.add(ename, eauth);
    modified = true;
    cout << "added entity " << ename << " auth " << eauth << std::endl;
  }
  if (caps_fn) {
    ConfFile cf;
    std::deque<std::string> parse_errors;
    if (cf.parse_file(caps_fn, &parse_errors) != 0) {
      cerr << "could not parse caps file " << caps_fn << std::endl;
      exit(1);
    }
    complain_about_parse_errors(&parse_errors);
    map<string, bufferlist> caps;
    const char *key_names[] = { "mon", "osd", "mds", NULL };
    for (int i=0; key_names[i]; i++) {
      std::string val;
      if (cf.read("global", key_names[i], val) == 0) {
        bufferlist bl;
        ::encode(val, bl);
        string s(key_names[i]);
        caps[s] = bl; 
      }
    }
    keyring.set_caps(ename, caps);
    modified = true;
  }
  if (caps.size()) {
    keyring.set_caps(ename, caps);
    modified = true;
  }
  if (set_auid) {
    keyring.set_uid(ename, auid);
    modified = true;
  }

  // read commands
  if (list) {
    keyring.print(cout);
  }
  if (print_key) {
    CryptoKey key;
    if (keyring.get_secret(ename, key)) {
      cout << key << std::endl;
    } else {
      cerr << "entity " << ename << " not found" << std::endl;
    }
  }

  // write result?
  if (modified) {
    bufferlist bl;
    if (bin_keyring) {
      ::encode(keyring, bl);
    } else {
      std::ostringstream os;
      keyring.print(os);
      string str = os.str();
      bl.append(str);
    }
    r = bl.write_file(fn, 0600);
    if (r < 0) {
      cerr << "could not write " << fn << std::endl;
    }
    //cout << "wrote " << bl.length() << " bytes to " << fn << std::endl;
  }

  return 0;
}
